gobblin-core/src/main/java/gobblin/writer/http/AsyncHttpWriter.java

```diff
@@ -0,0 +1,90 @@
+package gobblin.writer.http;
+
+import java.io.IOException;
+import java.util.Queue;
+
+import gobblin.async.DispatchException;
+import gobblin.http.HttpClient;
+import gobblin.http.ResponseHandler;
+import gobblin.http.ResponseStatus;
+import gobblin.writer.WriteResponse;
+
+
+/**
+ * This class is an {@link AsyncHttpWriter} that writes data in a batch, which
+ * is sent via http request
+ *
+ * @param <D> type of record
+ * @param <RQ> type of request
+ * @param <RP> type of response
+ */
+public class AsyncHttpWriter<D, RQ, RP> extends AbstractAsyncDataWriter<D> {
+  public static final int DEFAULT_MAX_ATTEMPTS = 3;
+
+  private final HttpClient<RQ, RP> client;
+  private final ResponseHandler<RP> responseHandler;
+  private final AsyncWriteRequestBuilder<D, RQ> requestBuilder;
+  private final int maxAttempts;
+
+  public AsyncHttpWriter(AsyncHttpWriterBaseBuilder builder) {
+    super(builder.getQueueCapacity());
+    this.client = builder.getClient();
+    this.requestBuilder = builder.getAsyncRequestBuilder();
+    this.responseHandler = builder.getResponseHandler();
+    this.maxAttempts = builder.getMaxAttempts();
+  }
+
+  @Override
+  protected void dispatch(Queue<BufferedRecord<D>> buffer) throws DispatchException {
+    AsyncWriteRequest<D, RQ> asyncWriteRequest = requestBuilder.buildRequest(buffer);
+    if (asyncWriteRequest == null) {
+      return;
+    }
+
+    RQ rawRequest = asyncWriteRequest.getRawRequest();
+    RP response;
+
+    int attempt = 0;
+    while (attempt < maxAttempts) {
+      try {
+        response = client.sendRequest(rawRequest);
+      } catch (IOException e) {
+        // Retry
+        attempt++;
+        if (attempt == maxAttempts) {
+          asyncWriteRequest.onFailure(e);
+          throw new DispatchException("Write failed on IOException", e);
+        } else {
+          continue;
+        }
+      }
+
+      ResponseStatus status = responseHandler.handleResponse(response);
```

Comment:
It might be better to include an enum response status type in `ResponseStatus`. This lets the responseHandler figure out what the response means (for example, and unexpected 2xx response code could mean we're sending to the wrong endpoint).